Documentation/design-docs/first-class-structs.md

```diff
@@ -15,6 +15,261 @@ Primary Objectives
+
+* On the RHS of an assignment
+  * The lhs provides the “shape” for an assignment. Note, however, that the LHS isn't always available
+    to optimizations, which has led to pessimization, e.g. 
+    [#23739 Block the hoisting of TYP_STRUCT rvalues in loop hoisting](https://github.com/dotnet/coreclr/pull/23739)
+
+* As a call argument
+  * In this context, it must be one of: `GT_OBJ`, `GT_LCL_VAR`, `GT_LCL_FLD` or `GT_FIELD_LIST`
+
+* As an operand to a hardware or SIMD intrinsic (for `TYP_SIMD*` only)
+  * In this case the struct handle is generally assumed to be unneeded, as it is captured (directly or
+    indirectly) in the `GT_SIMD` or `GT_HWIntrinsic` node.
+
+After morph, a struct-typed value on the RHS of assignment is one of:
+* `GT_IND`: in this case the LHS is expected to provide the struct handle
+  * Proposed: `GT_IND` would no longer be used for struct types
+* `GT_CALL`
+* `GT_RET_EXPR`
+* `GT_LCL_VAR`
+* `GT_LCL_FLD`
+  * Proposed: These would only be used for actual fields.
+* `GT_SIMD`
+* `GT_OBJ` nodes can also be used as rvalues when they are call arguments
+  * Proposed: `GT_OBJ` nodes can be used in any context where a struct rvalue or lvalue might occur,
+    except after morph when the struct is independently promoted.
+
+Struct IR Phase Transitions
+---------------------------
+
+There are three phases in the JIT that make changes to the representation of struct nodes and lclVars:
+
+* Importer
+  * Vector types are normalized to the appropriate `TYP_SIMD*` type. Other struct nodes have `TYP_STRUCT`.
+  * Proposed: all struct-valued nodes that are created with a class handle will retain an index into
+    the `ClassLayout` cache.
+  * Proposed: Also normalize to new types `TYP_STRUCT1`, etc.
+
+* Struct promotion
+  * Fields of promoted structs become separate lclVars (scalar promoted) with primitive types.
+    * This is currently an all-or-nothing choice (either all fields are promoted, or none), and is
+      constrained in the number of fields that can be promoted.
+    * Proposed: Support additional cases. See [Improve Struct Promotion](#Improve-Struct-Promotion)
+
+* Global morph
+  * Some promoted structs are forced to stack, and become “dependently promoted”.
+    * Proposed: promoted structs are forced to stack ONLY if address taken.
+      * This includes removing unnecessary pessimizations of block copies. See [Improve and Simplify Block Assignment Morphing](#Block-Assignments).
+
+  * Call args 
+    * If the struct has been promoted
+      * It is morphed to `GT_FIELD_LIST` if it is passed on the stack, or if it is passed
+        in registers that exactly match the types of its fields.
+        * Proposed: This transformation would be made even if it is passed in non-matching
+          registers. The necessary transformations for correct code generation would be
+          made in `Lowering`.
+      * If it is passed by reference, a copy is made, and that copy is passed according to
```

Comment:
I'm wondering if copies required by implicit by ref parameters shouldn't be done in lowering. It's not clear if there's any advtange in doing those copies earlier as the required temps are address exposed and thus ignored by optimizer for most purposes. Perhaps we can insert these copies after the liveness pass before LSRA - the last use information in produces should allow the elision of some copies.